core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1878,6 +1878,11 @@ && hasSortByOrdinal()) {
             && (hasNestedAgg || hasNestedWindowedAgg)) {
           return true;
         }
+        if (!dialect.supportsNestedAggregations()
+            && agg.getInput() instanceof Project
```

Comment:
What if input is Window object?

Reply:
And also `agg.getInput()` could be a `Calc` as well.

IMO something is deeply wrong with `needNewSubQuery` in general in terms of that it uses `instanceof Project` in multiple places, but does not really care about having `Calc` or `Window` on the same position, instead.

One of the possible solutions that could make whole code of RelToSqlConverter simpler, is to do some "normalization" of relational tree inside `SqlImplementor::visitRoot(...)` method in order to transform it to a form having less possible amount of node types.

But I guess that is something that needs to be discussed on the mailing list.

WDYT?

Reply:
When will the input of Aggreagte be Window?

Window only appear when you use `ProjectToWindowRule` and I think the window will not be input of Aggregate.

Reply:
@JiajunBernoulli 

It looks that you are right, at least I cannot reproduce an issue with `Window` under `Aggregate` that would also have a problem of aggregation by window expression.

But I guess that part of my comment about `Calc` is still relevant.